spec/ic.did

```diff
@@ -112,12 +112,10 @@ service ic : {
   http_request : (record {
     url : text;
     max_response_bytes: opt nat64;
-    method : variant { get; head; post };
+    http_method : variant { GET; HEAD; POST };
     headers: vec http_header;
     body : opt blob;
-    transform : opt variant {
-      function: func (http_response) -> (http_response) query
-    };
+    transform_method_name : opt text;
```

Comment:
Why are we changing this ? I will be on the side of using as much as type safety as possible .

Reply:
Indeed, this is not clear, could someone clarify?

Reply:
We will change the enumerations to lowercase in the implementation.
However, we wonder whether using "method" or "http_method" is better, considering that it is part of a record that has "httP" in its name.

Reply:
We need to analyze whether both Motoko and Rust can support this. What constraints does this impose on the use of potential other languages like C?

Reply:
@nomeata : Can you help in clarifying whether and how one would implement in Rust against the current specification (lines 118-120)? Does it preclude other languages with a less expressive type system like C to be used for developing canisters?
I would really like to keep the spec as in 118-120. Thanks for any input!

Reply:
Motoko has no problem supporting function references. On the Rust side, you can decode it to https://docs.rs/candid/0.7.16/candid/types/reference/struct.Func.html. But there is a problem in encoding the reference type. There are some workaround implemented in `ic-proxy`, but not ideal. See this issue: https://github.com/dfinity/candid/issues/273

Reply:
I'll defer to @chenyan-dfinity, he's thenl expert here :-)

Reply:
We had an extensive discussion on this internally with folks including @roman-kashitsyn, @Dfinity-Bjoern, and @dsarlis with the outcome that the "right" way of modeling this is analog to system functions like canister_init or canister_inspect_message as this best reflects the nature of a transform function, so we intend to change it to reflect that. This is in the interest of overall consistency with how things are modeled in the interface specification.

Reply:
I find it a very strange outcome that suddenly you are dropping to the Wasm level here.

The HTTP service is at a different level of abstraction as the low-level IC messaging protocol. `canister_init` and `canister_inspect_message` are clearly low-level canister interfaces; they are there to implement the Canister itself and build the IC inter-canister calling layer.

But the HTTP service could well be a separate canister! So it clearly is a protocol that can (and should) live on top of the lower-level, generic IC messaging service. The fact that the service’s implementation is magic and can’t be a normal canister does not affect how it should be interacted with, in my opinion.

So just like we reach all the management features not via additional Wasm-level entry points or System APIs, but by using IC-level calls, the same should be true for the HTTP component. 

And that means that that interface, including the “transform“ callback, should use the interface of ordinary inter-canister calls.

Reply:
That's also a perfectly valid view. One minor problem is that the HTTP adapter always calls into the replica on the same machine, assuming the originating canister exposes the method. The adapter should make the transform call through boundary nodes if we want an arbitrary callback.

Reply:
Is that an observable difference? Sounds like an implementation detail/optimization to me: A replica can of course handle a query call to a canister it “owns” directly, if it wants to.

It’s a bit like inter-canister calls between canisters on the same subnet being handled without involving cross-net messaging, because it would be kinda silly to have cross-net messaging between a subnet and itself. But it wouldn't be wrong, and the abstract interface does not mandate one or the other.

Reply:
> Is that an observable difference?

It will become observable if you specify a callback pointing into a canister on another subnet.

Reply:
Right, that's why the design so far had the extra requirement that the callback has to be to the same canister, don't it?

I admit it's a bit of a odd restriction, but much less if a severe layering violation than dropping to the wasm level here.

And with this restriction, it's not observable anymore, and the replica is welcome to resolve it locally.

Reply:
Yes, it's essential for the trust model of the feature that the transform function is executed on the same replica that makes the network request.

Reply:
> Right, that's why the design so far had the extra requirement that the callback has to be to the same canister, don't it?

I'm somewhat confused. If there was such a restriction on the design, then what's the point of providing a generic callback that would allow someone to call into any other canister (on the interface level at least)? Of course, the implementation can provide restrictions on who can be called (i.e. only resolve locally the request if you can) but then the interface design seems wrong to me.

Given the whole direction on the design of this feature, the choice of a system method seems the most natural to capture that:
1. This transform should only be ever called on the canister that made the original http request.
2. There's no sender, because we're not making a generic query call, but rather the system needs to invoke a special function on a canister at some point in time.

Reply:
The point of using a Candid function reference is
 * to get strong typing
 * to not have to hard code the name of the query 

The first is the interesting one, the second a nice to have.

Adding wasm layer entry points doesn't really scale. Imagine the churn it every new management canister feature was a wasm level system API function! This is the same here, I'd say.

Candid function references aren't use a lot yet, but they are there in the design for the IC programming model, and I think it's a step in the right direction to start using them.


If they are really that unwanted for some reason, at least use a normal query method with a well-known name, as in an earlier iteration of the adaptor design. That at least doesn't open up the abstract idea of what a canister is needlessly.

Reply:
> to get strong typing

Yeah, in Motoko. If you write in Rust, it doesn't provide you with an extra benefit with the current library support.

> Adding wasm layer entry points doesn't really scale. Imagine the churn it every new management canister feature was a wasm level system API function! This is the same here, I'd say.

I agree with this point in general and I don't think we've been adding wasm entry points left and right -- we're generally cautious on when new wasm entry points are introduced.

In this case, it feels to me like the system method approach captures the intent better as I explained in my previous comment. My understanding is that we don't need a normal query method that's generally available to anyone, we need some method that the system (and the system only) can invoke when it needs to sanitise the http response.

> That at least doesn't open up the abstract idea of what a canister is needlessly.

I'm not sure I understand your point here. System methods are well established in the interface spec, we're not trying to introduce something completely new and unique.

Reply:
> Yeah, in Motoko. If you write in Rust, it doesn't provide you with an extra benefit with the current library support.

Exactly - about time to create more incentives to improve the Rust tooling with regards to inter-canister type safety :-)

> and the system only

Good point! The design with the function reference would be forward compatible to the world with real closures, a.k.a. capabilities, where the canister can generate a function reference that really can only be called by those that have it. Not sure if we ever get that, but it has been part if the vision since the beginning. So another argument for doing it this way :-)

The interface spec makes a clear abstract around the low level system API, in https://internetcomputer.org/docs/current/references/ic-interface-spec/#abstract-canisters, and then has a _separare_ section of how that maps to Wasm. Keeping that interface narrow and to the point is, I believe, useful to keep the system complexity manageable. And any feature that can be added without touching that interface should be added that way, unless there are really strong reasons to break that abstraction (e.g. a really bad performance hit).

We only have wasm canisters and that will probably stay that way, but I think a good design guidance is to imagine we had many different low level canister implementations (Wasm, Java Bytecode, Lua…), all with their own system API, but all implementing this abstract canister interface. How would we do it then? Very likely we'd stay above that interface, so that we don't have to change n different system APIs.

Reply:
BTW, I think we are debating exactly the same points as when we debated whether to use a virtual canister for management, or use system API. I had a hard time convincing Johan of the virtual canister approach, and only @ielashi managed to frame it in a way that convinced him. Maybe Islam can help coming to a conclusion here too :-)

Reply:
I am feeling the pressure now :-) I'm catching up here, so apologies in advance if I'm missing some major points.

I think we're all in agreement that we should keep the low level system API as minimal as possible, and the dilemma here is whether the transform method is a primitive that needs to be added to the system API, or whether it can be expressed using other primitives.

If I understand correctly, what makes the transform special, as Dimitris already mentioned, is the following:

1. There's no sender (and hence it's somewhat different from a query).
2. It should (ideally) not be invoked by any canister apart from the canister itself and the management canister.

IMHO, the first property isn't really necessary, and the second property can be enforced with other lower-level primitives.

Regarding "there's no sender", why is that the case? The canister is requesting from the management canister to do an HTTP request, and the _management canister_ then calls the transform method, so it seems perfectly consistent to have the management canister as the sender. Are there any reasons we need this special casing?

As for access control, which is more of a nice-to-have than a hard requirement, I believe we're also all in agreement that we should charge for queries at some point, and with that we'll need some mechanism to allow canisters to accept/reject processing queries (i.e. some kind of equivalent to `canister_inspect_message`). Once that primitive is available, then a canister can easily ensure that the transform method can only be called by itself and/or the management canister if it wishes.

With the above in mind, I'd be inclined to think the transform method is really no different than a query, and that it would be simpler to treat it as such rather than introduce a new primitive in the protocol.

As for the debate of whether to use a `func` or a `text` for the transform method, I don't have a strong opinion. I agree that `func` feels a bit too generic, but on the other hand I like the additional type safety that it can provide with the right support from a langauge's CDK, so I'd have a slight preference for that.

Reply:
I feel there's still a misunderstanding here on what we mean when we say the management canister "calls" the transform function. 

The design is that the _replica_ calls the transform function and not really the management canister. If it was an actual call from the management canister, I'd expect it to be similar to an inter-canister call but the feature is designed in a way that the call does not resemble an inter-canister call. Perhaps it could have been designed and implemented in such a way but I simply don't see it in the current design and implementation.

That said, I feel we're just going in circles here. This discussion was brought up internally and a few people (tagged here and including myself) found the system method approach to be matching better. It seems there's another set of people that feel differently and we probably need to agree we disagree. Each side has presented their arguments, so I propose we leave it up to @Dfinity-Bjoern and other spec owners (and ultimately PR approvers) to set the direction.

Reply:
> The design is that the _replica_ calls the transform function and not really the management canister.

That sentence doesn’t make sense to me. The management canister isn’t a think, it is an abstract concept, a facade _for the replica_ (or the subnet, if you want). Just like if a canister calls the management canister to get some randomness, of course what’s happening really is the replica handles that.

But yes, I think the arguments are on the table. Poor @Dfinity-Bjoern will have to make a call :-)

Reply:
I must admit @nomeata swayed me to the let-it-be-a-query direction. The main reason reason is that, from a canister perspective, the API of calling `aaaaa-aa` and then getting a "callback" (in quotes since it's not a callback in the interface spec sense) from `aaaaa-aa` makes a lot of sense. It is, in a sense, upward-compatible to providing the functionality through an actual canister. So if I have to make a call, then we go with:

- we leave it as a query that is invoked (with caller `aaaaa-aa`),
- we also keep the type-safe variant.